File: LeetcodeQuestions/leetcode_challenges/April_challenge/letter_combinations_of_a_phone_number.rb
def letter_combinations(digits)
    combos = []
    return combos if !digits.length
    mapping = { 2=> "abc", 3=> "def", 4=> "ghi", 
                   5=> "jkl", 6=> "mno", 7=> "pqrs", 
                   8=> "tuv", 9=> "wxyz"}
    letter_combo_recur(0, "", combos, digits, mapping)
    combos;
end

def letter_combo_recur(idx, path, combos, digits, mapping)
    if idx == digits.length
        combos.push(path)
        return
    end
    letters = mapping[digits[idx].to_i]
    (0...letters.length).each do |i|
        char = letters[i]
        letter_combo_recur(idx + 1, path + char, combos, digits, mapping)
    end
end

p(letter_combinations("23"))
p(letter_combinations("234"))

File: LeetcodeQuestions/leetcode_challenges/April_challenge/letter_combinations_of_a_phone_number.py
def letter_combinations(digits):
    combos = []
    if not len(digits):
        return combos
    mapping = {2: "abc", 3: "def", 4: "ghi",
               5: "jkl", 6: "mno", 7: "pqrs",
               8: "tuv", 9: "wxyz"}
    letter_combo_recur(0, "", digits, combos, mapping)
    return combos


def letter_combo_recur(idx, path, digits, combos, mapping):
    if idx == len(digits):
        combos.append(path)
        return
    letters = mapping[int(digits[idx])]
    for i in range(len(letters)):
        char = letters[i]
        letter_combo_recur(idx + 1, path + char, digits, combos, mapping)


print(letter_combinations("23"))
print(letter_combinations("234"))
